pkg/codegen/docs/gen_method.go

```diff
@@ -0,0 +1,329 @@
+	var params []formalParam
+
+	params = append(params, formalParam{
+		Name: "self",
+	})
+
+	if f.Inputs != nil {
+		// Filter out the __self__ argument from the inputs.
+		args := make([]*schema.Property, 0, len(f.Inputs.InputShape.Properties)-1)
+		for _, arg := range f.Inputs.InputShape.Properties {
+			if arg.Name == "__self__" {
+				continue
+			}
+			args = append(args, arg)
+		}
+		// Sort required args first.
+		sort.Slice(args, func(i, j int) bool {
+			pi, pj := args[i], args[j]
+			switch {
+			case pi.IsRequired() != pj.IsRequired():
+				return pi.IsRequired() && !pj.IsRequired()
+			default:
+				return pi.Name < pj.Name
+			}
+		})
+		for _, arg := range args {
+			typ := docLanguageHelper.GetLanguageTypeString(mod.pkg, mod.mod, arg.Type, true /*input*/)
+			var defaultValue string
+			if !arg.IsRequired() {
+				defaultValue = " = None"
+			}
+			params = append(params, formalParam{
+				Name:         python.PyName(arg.Name),
+				DefaultValue: defaultValue,
+				Type: propertyType{
+					Name: typ,
+				},
+			})
+		}
+	}
+	return params
+}
+
+// genMethodArgs generates the arguments string for a given method that can be
+// rendered directly into a template. An empty string indicates no args.
+func (mod *modContext) genMethodArgs(r *schema.Resource, m *schema.Method,
+	methodNameMap map[string]string) map[string]string {
+
+	f := m.Function
+
+	functionParams := make(map[string]string)
+	for _, lang := range supportedLanguages {
+		var (
+			paramTemplate string
+			params        []formalParam
+		)
+		b := &bytes.Buffer{}
+
+		paramSeparatorTemplate := "param_separator"
+		ps := paramSeparator{}
+
+		var hasArgs bool
+		optionalArgs := true
+		if f.Inputs != nil {
+			for _, arg := range f.Inputs.InputShape.Properties {
+				if arg.Name == "__self__" {
+					continue
+				}
+				hasArgs = true
+				if arg.IsRequired() {
+					optionalArgs = false
+				}
+			}
+		}
+
+		if !hasArgs {
+			functionParams[lang] = ""
+			continue
+		}
+
+		switch lang {
+		case "nodejs":
+			params = mod.genMethodTS(f, resourceName(r), methodNameMap["nodejs"], optionalArgs)
+			paramTemplate = "ts_formal_param"
+		case "go":
+			params = mod.genMethodGo(f, resourceName(r), methodNameMap["go"], optionalArgs)
+			paramTemplate = "go_formal_param"
+		case "csharp":
+			params = mod.genMethodCS(f, resourceName(r), methodNameMap["csharp"], optionalArgs)
+			paramTemplate = "csharp_formal_param"
+		case "python":
+			params = mod.genMethodPython(f)
+			paramTemplate = "py_formal_param"
+			paramSeparatorTemplate = "py_param_separator"
+
+			docHelper := getLanguageDocHelper(lang)
+			methodName := docHelper.GetMethodName(m)
+			ps = paramSeparator{Indent: strings.Repeat(" ", len("def (")+len(methodName))}
+		}
+
+		n := len(params)
+		if n == 0 {
+			functionParams[lang] = ""
+			continue
+		}
+
+		for i, p := range params {
+			if err := templates.ExecuteTemplate(b, paramTemplate, p); err != nil {
+				panic(err)
```

Comment:
Nit: could use normal err handling here?

Reply:
I agree with the nitpick, and normally I would prefer the standard err handling, but this is patterned off of what is [done](https://github.com/pulumi/pulumi/blob/85c15b05e4b3b8f97375211209d3b1507a31d9fe/pkg/codegen/docs/gen.go#L1032-L1041) [elsewhere](https://github.com/pulumi/pulumi/blob/85c15b05e4b3b8f97375211209d3b1507a31d9fe/pkg/codegen/docs/gen.go#L1260-L1269) [in](https://github.com/pulumi/pulumi/blob/85c15b05e4b3b8f97375211209d3b1507a31d9fe/pkg/codegen/docs/gen_function.go#L269-L278) this package, and replumbing it to do the standard thing would be a larger change than I'd like to bite off right now. So I'm going to keep it as-is.

Reply:
Right, there's probably a reason for it (like the errors being almost impossible).